File: soglia1848/native/nsa.cs
using System; using System.IO; using System.Linq; using System.Collections.Generic;
class N {
  static float[] LP; static int[] seq; static int ntypes; static int[][] cand; static double lam;
  static double Score(int[] m){
    double t=0; int h=0; int n=0;
    for(int i=0;i<seq.Length;i++){
      int k=seq[i];
      if(k<0){ int idx=h*27; t+=LP[idx]; h=idx%531441; n++; continue; }
      int[] s=cand[m[k]];
      for(int j=0;j<s.Length;j++){ int idx=h*27+s[j]; t+=LP[idx]; h=idx%531441; n++; }
    }
    t+=LP[h*27];
    return t+lam*n;
  }
  static void Main(string[] a){
    // args: lm.bin seq.txt cand.txt seed iters restarts T0 T1 lam out pletter
    byte[] b=File.ReadAllBytes(a[0]); LP=new float[b.Length/4]; Buffer.BlockCopy(b,0,LP,0,b.Length);
    var lines=File.ReadAllLines(a[1]);
    ntypes=int.Parse(lines[0]); seq=lines[1].Split(' ').Select(int.Parse).ToArray();
    int[] fixedv=lines.Length>2 && lines[2].Trim().Length>0 ? lines[2].Split(' ').Select(int.Parse).ToArray() : Enumerable.Repeat(-1,ntypes).ToArray();
    var cl=File.ReadAllLines(a[2]).Where(x=>x.Length>0).ToArray();
    cand=cl.Select(x=>x.Select(ch=>ch-'a'+1).ToArray()).ToArray();
    int nletters=cl.Count(x=>x.Length==1);
    int seed=int.Parse(a[3]); long iters=long.Parse(a[4]); int R=int.Parse(a[5]);
    double T0=double.Parse(a[6]),T1=double.Parse(a[7]); lam=double.Parse(a[8]);
    double pl=double.Parse(a[10]);
    var rnd=new Random(seed); var outw=new StreamWriter(a[9]);
    for(int r=0;r<R;r++){
      int[] m=new int[ntypes];
      for(int i=0;i<ntypes;i++) m[i]= fixedv[i]>=0? fixedv[i] : rnd.Next(nletters);
      double cur=Score(m), best=cur; int[] bm=(int[])m.Clone();
      for(long it=0;it<iters;it++){
        double T=T0*Math.Pow(T1/T0,(double)it/iters);
        int k=rnd.Next(ntypes); if(fixedv[k]>=0) continue;
        int old=m[k],k2=-1,old2=0;
        double u=rnd.NextDouble();
        if(u<0.6) m[k]= rnd.NextDouble()<pl ? rnd.Next(nletters) : rnd.Next(cand.Length);
        else { k2=rnd.Next(ntypes); if(fixedv[k2]>=0) continue; old2=m[k2]; m[k]=old2; m[k2]=old; }
        double nw=Score(m);
        if(nw>=cur || rnd.NextDouble()<Math.Exp((nw-cur)/T)){ cur=nw; if(cur>best){best=cur; Array.Copy(m,bm,ntypes);} }
        else { m[k]=old; if(k2>=0) m[k2]=old2; }
      }
      outw.WriteLine(best.ToString("F2")+"\t"+string.Join(" ",bm)); outw.Flush();
    }
    outw.Close();
  }
}

File: soglia1848/nsa.py
import numpy as np, re, sys, math, collections
from numba import njit
A=27
LP=np.load(sys.argv[1] if len(sys.argv)>1 and sys.argv[1].endswith('.npy') else 'lm5sp.npy')

def load_units(fn='ct2.txt'):
    s=open(fn).read().strip()
    out=[]
    for r in re.split('5',s):
        if len(r)%2==0: out+=[r[i:i+2] for i in range(0,len(r),2)]
        else: out+=[r[i:i+2] for i in range(0,len(r)-1,2)]+['L'+r[-1]]
        out.append('_')
    o=[]
    for u in out:
        if u=='_' and (not o or o[-1]=='_'): continue
        o.append(u)
    return o

@njit(cache=True)
def score(seq,m,LP):
    tot=0.0
    h0=0;h1=0;h2=0;h3=0
    for k in seq:
        c=0 if k<0 else m[k]
        if c<0: continue   # null
        tot+=LP[(((h0*27+h1)*27+h2)*27+h3)*27+c]
        h0=h1;h1=h2;h2=h3;h3=c
    tot+=LP[(((h0*27+h1)*27+h2)*27+h3)*27+0]
    return tot

@njit(cache=True)
def anneal(seq,m,LP,fixed,iters,T0,T1,seed,allowed):
    np.random.seed(seed)
    n=len(m)
    cur=score(seq,m,LP)
    best=cur;bm=m.copy()
    na=len(allowed)
    for i in range(iters):
        T=T0*(T1/T0)**(i/iters)
        k=np.random.randint(n)
        if fixed[k]: continue
        old=m[k]
        if np.random.random()<0.8:
            m[k]=allowed[np.random.randint(na)]
            k2=-1
        else:
            k2=np.random.randint(n)
            if fixed[k2]: continue
            old2=m[k2]; m[k]=old2; m[k2]=old
        new=score(seq,m,LP)
        if new>=cur or np.random.random()<math.exp((new-cur)/T):
            cur=new
            if cur>best: best=cur;bm[:]=m
        else:
            m[k]=old
            if k2>=0: m[k2]=old2
    return best,bm

def show(seq,m,types):
    return ''.join(' ' if k<0 else ('' if m[k]<0 else chr(96+m[k])) for k in seq)
